File: 250721/안전 지대/comfort-zone.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const grid = input.slice(1, 1 + Number(n)).map(line => line.split(' ').map(Number));

const answer = []
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];
function dfs(x, y, visited, k){
for(let i = 0; i<4; i++){
    const nx = x + dx[i];
    const ny = y + dy[i];

    if(0<= nx && nx < m && 0 <= ny && ny < n && !visited[ny][nx] && grid[ny][nx] > k){
        visited[ny][nx] = true;
        dfs(nx, ny, visited, k)
    }
}
}

for(let k = 1; k <= 100; k ++){
    const visited = Array(n).fill(false).map(() => Array(m).fill(false));
    let count = 0
    for(let i = 0; i < n; i++){
        for(let j = 0; j < m; j++){
            if(!visited[i][j] && grid[i][j] > k){
                visited[i][j] = true;
                dfs(j, i, visited, k)
                count++;
            }
        }
    }
    if(Math.max(...answer.map(val => val[1])) <= count){
        answer.push([k, count])
    }
      
}
if (answer.length > 0) {
    const maxSafePlace = Math.max(...answer.map(val => val[1]));
    const mink = answer.filter(val => val[1] === maxSafePlace).sort((a, b) => a[0] - b[0])[0][0];
    console.log(mink + ' ' + maxSafePlace);
} else {
    console.log("1 0");
}

File: 250820/안전 지대/comfort-zone.py
import sys
sys.setrecursionlimit(10**6)

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]



dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def dfs(x, y, k, visited):
    for i in range(4):
        nx = dx[i] + x
        ny = dy[i] + y

        if 0 <= nx and nx < m and 0 <= ny and ny < n and grid[ny][nx] > k and not visited[ny][nx]:
            visited[ny][nx] = True
            dfs(nx, ny, k, visited)

max_cnt = 0
ans_k = 1
for k in range(1, 101):
    visited = [[False] * m for _ in range(n)]
    cnt = 0
    for i in range(n):
        for j in range(m):
            if grid[i][j] > k and not visited[i][j]:
                visited[i][j] = True
                cnt+=1
                dfs(j, i, k, visited)

    if max_cnt < cnt:
        ans_k = k
        max_cnt = cnt

print(ans_k, max_cnt)
